Trame Dashboard

proy30 opened the pull request:
Added a trame dashboard.

proy30 committed:
Added trame dashboard

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
Added 'variable referencing' tab in lattice settings

proy30 committed:
Fix merge conflict

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
Imported elements module for proper simulation running

elements module was not loading in, resulting in errors to read latticeElements

proy30 committed:
Merge branch 'development' of github.com:proy30/impactx into development

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
Labeled static methods in analyzeFunctions

proy30 committed:
Merge branch 'development' of github.com:proy30/impactx into development

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
Added static methods across all files which need

This helps code readability

proy30 committed:
fix: fixed CodeQL warning

avoided using return value of a procedure that returns None

proy30 committed:
fix: ensuring itrable input in generalFunctions.py

Added check

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
Fixed minor CodeQL errors

proy30 committed:
Added base functionality to print in terminal view

ax3l commented on src/python/impactx/dashboard/requirements.txt:
```diff
@@ -0,0 +1,45 @@
+aiohttp==3.9.5
```
Usually we use compatible `~=` or `>=` here to have a wider solution space for package management (conda, spack, pip, etc.). We seldomly can use the _exact_ same versions on various platforms we deploy to.

ax3l commented on src/python/impactx/dashboard/output_distribution_parameters.txt:
```diff
@@ -0,0 +1,11 @@
+distr = distribution.Waterbag(
+    lambdaX=None,
+    lambdaY=None,
```
Did you mean to exclude this generated file in our base `.gitignore` file?

ax3l commented on src/python/impactx/dashboard/output_latticeElements_parameters.txt:
```diff
@@ -0,0 +1,4 @@
+latticeElements = [
```
Did you mean to exclude this generated file in our base `.gitignore` file?

ax3l commented on src/python/impactx/dashboard/main.py:
```diff
@@ -0,0 +1,128 @@
+        return application()
+
+
+application()
```
Does this belong here (import time) or into `__main__` time?

Note that we can create an entry point to start this in our `setup.py` like this:
https://github.com/openPMD/openPMD-api/blob/4f83a7e91f7a841eaee741e0dc0729625e0b11f5/setup.py#L196-L202
https://github.com/openPMD/openPMD-api/blob/4f83a7e91f7a841eaee741e0dc0729625e0b11f5/src/binding/python/openpmd_api/ls/__main__.py
Let's add that and call it `impactx-dashboard`?

proy30 replied:
added entry point

ax3l commented on src/python/impactx/dashboard/Toolbar/exportTemplate.py:
```diff
@@ -0,0 +1,72 @@
+
+def retrieve_state_content():
+    """
+    This function builds the output file based on user inputs
```
You mean the input file?
```suggestion
    This function builds the input file based on user inputs

```

proy30 replied:
Since this function is for exporting user inputs,  should it be "builds the 'export'" instead? When I think of input file, I am thinking of a file the user would provide to the system, but this function is meant for the other way around.

ax3l replied:
Yes, that is more explicit. Please change as you suggested :)

ax3l commented on src/python/impactx/dashboard/Toolbar/exportTemplate.py:
```diff
@@ -0,0 +1,72 @@
+beam.units = static
+beam.kin_energy = {state.kin_energy_MeV}
+beam.charge = {state.charge}
+beam.particle = electron
```
Why do we write this file (AMReX inputs file for `amrex::ParmParse`) when we use the Python API of ImpactX to set up a simulation?

Is this for export?
Do we want to instead export a Python script of the run (that includes the lattice, too)?

ax3l commented on src/python/impactx/dashboard/Run/optimize_triplet/run_triplet.py:
```diff
@@ -0,0 +1,256 @@
+# -----------------------------------------------------------------------------
```
It appears that all files in the `Run/` folder are auto-generated and should be added to `.gitignore`?

ax3l commented on src/python/impactx/dashboard/Optimize/optimizeMain.py:
```diff
@@ -0,0 +1,54 @@
+# -----------------------------------------------------------------------------
+
+
+class Optimize:
```
Can you please add doc strings to the class?
```suggestion
class Optimize:
    """..."""
```

ax3l commented on src/python/impactx/dashboard/Input/trameFunctions.py:
```diff
@@ -0,0 +1,52 @@
+}
+
+
+class trameFunctions:
```
Please add doc strings to this class: Where is this class used, what high-level GUI element does it build.

ax3l commented on src/python/impactx/dashboard/Input/inputParametersCard/inputMain.py:
```diff
@@ -0,0 +1,153 @@
+# -----------------------------------------------------------------------------
+
+
+class inputParameters:
```
Please add doc strings to this class: Where is this class used, what high-level GUI element does it build.

ax3l commented on src/python/impactx/dashboard/Input/generalFunctions.py:
```diff
@@ -0,0 +1,254 @@
+ANSI_RESET = "\033[0m"
+
+
+class generalFunctions:
```
Please add doc strings to this class: Where is this class used, what high-level GUI element does it build.

ax3l commented on src/python/impactx/dashboard/Input/distributionParametersCard/twissParameters.txt:
```diff
@@ -0,0 +1,88 @@
+twiss_parameters = {
```
Is this file auto-generated and should be added to `.gitignore`?

ax3l commented on src/python/impactx/dashboard/Input/distributionParametersCard/distributionMain.py:
```diff
@@ -0,0 +1,203 @@
+# -----------------------------------------------------------------------------
+
+
+class distributionParameters:
```
Please add doc strings to this class: Where is this class used, what high-level GUI element does it build.

ax3l commented on src/python/impactx/dashboard/Input/distributionParametersCard/distributionFunctions.py:
```diff
@@ -0,0 +1,11 @@
+class distributionFunctions:
```
Please add doc strings to this class: Where is this class used, what high-level GUI element does it build/validate.

ax3l commented on src/python/impactx/dashboard/Input/distributionParametersCard/distributionFunctions.py:
```diff
@@ -0,0 +1,11 @@
+
+    @staticmethod
+    def classAndParametersAndDefaultValueAndType_Twiss():
+        file_path = "/mnt/c/Users/parth/Downloads/vsCode/TrameApplication/Input/distributionParametersCard/twissParameters.txt"
```
please generalize the hard-coded path

ax3l commented on src/python/impactx/dashboard/Input/distributionParametersCard/distributionFunctions.py:
```diff
@@ -0,0 +1,11 @@
+        with open(file_path, "r") as file:
+            content = file.read()
+            local_vars = {}
+            exec(content, {}, local_vars)
```
Do we need to pass those parameters as a file or could we just pass a dictionary argument between functions instead?

ax3l commented on src/python/impactx/dashboard/Input/Visualiztion/twiss_phase_space_ellipse/x_px.py:
```diff
@@ -0,0 +1,219 @@
+# -----------------------------------------------------------------------------
+
+
+class visualizeTwiss:
```
Please add doc strings to this class: Where is this class used, what high-level GUI element does it build.

ax3l commented on src/python/impactx/dashboard/Analyze/plotsMain.py:
```diff
@@ -0,0 +1,172 @@
+# -----------------------------------------------------------------------------
+
+
+class AnalyzeSimulation:
```
Please add doc strings to this class: Where is this class used, what high-level GUI element does it build.

ax3l commented on src/python/impactx/dashboard/Analyze/plot_over_s/overS.py:
```diff
@@ -0,0 +1,32 @@
+import plotly.graph_objects as go
+
+
+def line_plot_1d(selected_headers, filtered_data):
```
Please add a docstring with parameter descriptions :)

ax3l commented on src/python/impactx/dashboard/Analyze/analyzeFunctions.py:
```diff
@@ -0,0 +1,103 @@
+        file_path = distribution_parameters_file_path
+        safe_env = {"distribution": distribution, "distr": None}
+
+        with open(file_path, "r") as file:
+            exec(file.read(), safe_env)
```
Do we need to pass as a file or can we pass a dictionary argument?

proy30 committed:
Added files to .gitignore for Trame Dashboard

Includes the diag files that are generated by simulations, and two temporary files for beam distribution and lattice elements.

proy30 committed:
Minor format update and remove obsolete file now in .gitignore

proy30 committed:
Nearly completed terminal integration

Upon clicking the run simulation button, terminal output correctly prints in the terminal view. However, the simulation script is directly called from the terminal view, which is unallowing itself to retrieve the user-inputted valued. Right now, the terminal correctly works only if fixed valued are placed in the script.

proy30 committed:
Updated Trame Dashboard .gitignore

proy30 committed:
Removed file reading for beam distribution and lattice elements

Basic testing with this change. Functionality looks to be the same thus far.

proy30 committed:
Removed 'workflow' options

Implemented as wrong interpretation. Clicking 'run simulation' now runs the dataFrame example.

proy30 committed:
Added entry point to impactx-dashboard

User needs to install all impactx packages, and from there can run the trame application from the impactx directory with the 'impactx-dashboard' commandline

ax3l commented on src/python/impactx/dashboard/requirements.txt:
```diff
@@ -0,0 +1,45 @@
+six==1.16.0
+tenacity==8.3.0
+toolz==0.12.1
+trame==3.6.2
```
Please reduce the entries here to direct dependencies we need to install, e.g., `trame` and other packages you directly use, and their minimal versions.

Transitive dependencies are maintained by our dependencies, let us not list them here for maintainability.

proy30 replied:
Should the version be included as well?

ax3l replied:
Minimal known working version would be good, yes

proy30 committed:
Added __init__ files

proy30 committed:
Created entry point

proy30 committed:
File references are relative paths

proy30 committed:
Deletion of main.py (part of file rename)

proy30 committed:
Updated .gitignore

proy30 committed:
Updated file path reference for plotting over position plot

proy30 committed:
distribution and lattice variables call another function

Part of an old commit which removed the use of text files for distirbution/latticeelemtnt

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
Renamed class names

To follow standard Python convention

proy30 committed:
fixing pre-commit.ci errors

proy30 committed:
Removed unused code in simulation

proy30 committed:
Fixed convention and updated docstring/function names for DistributionParameters class

proy30 committed:
Updated docstrings

proy30 committed:
Fixed function convention spacing

proy30 committed:
Updated function conventions/renamed

proy30 committed:
Renamed files

proy30 committed:
Deleted/removed unused files, some format update

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
Fixing pre-commit.ci

proy30 committed:
Corrected f-string formatting

proy30 committed:
update .gitignore

proy30 committed:
moved main() to own file, jupyter-specific code in own file

proy30 committed:
Added class docstrings

proy30 committed:
Updated requirements.txt

Contains higher-level libraries needed to install, rather than including it's sublibraries

proy30 committed:
Modified code to follow Python/ImpactX styling and convention

proy30 committed:
Merge branch 'development' of github.com:proy30/impactx into development

proy30 committed:
Fixing some more conventions

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
Allow phase space projections to display

proy30 committed:
Squash, Adding originalPhaseSpaceProjections plot to Trame

For testing and code change monitoring purposes. Current matplotlib implementation is not quite compatible with the Trame Application.

proy30 committed:
Squash, removed histogram from originalPlot and changed to scatter

proy30 committed:
Squash

proy30 committed:
Updated phase space projection

proy30 committed:
Disabled not fully completed functionalities

proy30 committed:
fixed class naming convention

proy30 committed:
Merge branch 'development' of github.com:proy30/impactx into development

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l commented on src/python/impactx/dashboard/Toolbar/toolbarMain.py:
```diff
@@ -0,0 +1,190 @@
+from ..Input.trameFunctions import TrameFunctions
+from .exportTemplate import retrieve_state_content
+
+# -----------------------------------------------------------------------------
+# Trame setup
+# -----------------------------------------------------------------------------
+
+server = get_server(client_type="vue2")
+state, ctrl = server.state, server.controller
+
+state.selectedVisualization = None
```
Move this to a central import and call it as a function in all here?
```suggestion
# -----------------------------------------------------------------------------
# Trame setup
# -----------------------------------------------------------------------------

server, state, ctrl = init_trame_server()

state.selectedVisualization = None
```

proy30 committed:
Added trame_setup.py

ax3l commented on src/python/impactx/dashboard/Input/distributionParameters/distributionFunctions.py:
```diff
@@ -0,0 +1,20 @@
+
+        file_path = "/mnt/c/Users/parth/Downloads/vsCode/TrameApplication/Input/distributionParametersCard/twissParameters.txt"
+
+        with open(file_path, "r") as file:
+            content = file.read()
+            local_vars = {}
```
When I start up the current branch, the dashboard startup fails here.

proy30 committed:
Updated requirements.txt

proy30 committed:
Added ability to switch between png and interactive for phase space

proy30 committed:
Updated gitignore

proy30 committed:
Disabled non-completed features

proy30 committed:
Added initial doxygen

proy30 committed:
Added xterm with broken functionality

proy30 committed:
Merge branch 'development' of github.com:proy30/impactx into development

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
Added xterm library

proy30 committed:
Merge branch 'development' of github.com:proy30/impactx into development

proy30 committed:
xTerm init function created

proy30 committed:
Fixed simulation code

proy30 committed:
Added simple integration of simulation progress in terminal

proy30 committed:
Attempt at fixing loading diags file

proy30 committed:
Commented out unused toolbar elements

proy30 committed:
delete code for  unused toolbar elements

proy30 committed:
Removed commented out code and fixed some format

proy30 committed:
Modified doxygen for dashboard

proy30 committed:
Updated minor bugs

proy30 committed:
Removed code that is not used

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
MPI support in phase_space

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
MPI support in phase_space

proy30 committed:
Merge branch 'development' of github.com:proy30/impactx into development

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l committed:
Doc: Update Dashboard

- add to table of contents
- restructure users vs. developers
- do not check in a PNG, link to gist

ax3l committed:
Jupyter: Simplify/Rename Imports

ax3l committed:
Update Screenshot

ax3l committed:
Run `dos2unix` on all files

ax3l committed:
Docs: Developer Module

ax3l commented on src/python/impactx/dashboard/Input/generalFunctions.py:
```diff
@@ -0,0 +1,273 @@
+                return ["Must be an integer"]
+            try:
+                value = int(input_value)
+                if value < -10000000000:
+                    return ["Must be positive"]
```
This looks confusing to me. Should this read
```suggestion
                if value < 0:
                    return ["Must be positive"]
```
?

proy30 replied:
Yes, we need to fix this. I remember intentionally changing '<0" to '<-10000000000' before. I've seen that some parameters can be negative values (ie. k parameter in Quad element), and having the GUI not allow negative values was an issue. In a future PR, we can focus on setting boundaries for specific parameters?

ax3l replied:
Sounds good! Yes, let's remove this for now, strict positive values depend a lot on the exact parameter and quite a few allow negative values, too.

ax3l commented on src/python/impactx/dashboard/Input/generalFunctions.py:
```diff
@@ -0,0 +1,273 @@
+            if input_value is None:
+                return ["Must be a float"]
+            try:
+                value = float(input_value)
+                if value < -1000000000:
+                    return ["Must be positive"]
```
Same here?
```suggestion
                value = float(input_value)
                if value < 0:
                    return ["Must be positive"]
```

ax3l commented on src/python/impactx/dashboard/Input/generalFunctions.py:
```diff
@@ -0,0 +1,273 @@
+            error_details.append("LatticeListIsEmpty")
+
+        # Print all collected error messages
+        """
+        if error_details:
+            print("Errors detected in the following parameters:")
+            for error in error_details:
+                print(ANSI_RED + error + ANSI_RESET)
+        """
```
Does this print or is this a multi-line string that does nothing?

proy30 replied:
Debug code. Can remove.

ax3l commented on src/python/impactx/dashboard/Input/inputParameters/inputMain.py:
```diff
@@ -0,0 +1,160 @@
+        state.kin_energy_MeV = InputFunctions.value_of_kin_energy_MeV(
+            float(state.kin_energy), new_unit
+        )
+        # print(f"Value of - state.kin_energy_MeV (back-end value): {state.kin_energy_MeV}")
+        # print(f"Value of - state.kin_energy (front-end value): {state.kin_energy}")
+
+        # print(f"Units were changed to {new_unit}")
+    # print(f"old unit is {old_unit}")
+    # print(f"new unit is {new_unit}")
```
Old debug code, let's clean it up:
```suggestion
```

ax3l commented on src/python/impactx/dashboard/Input/inputParameters/inputMain.py:
```diff
@@ -0,0 +1,160 @@
+        state.npart_validation = []
+        state.kin_energy_validation = []
+        state.bunch_charge_C_validation = []
+
+        # print(f"Initial value of - state.kin_energy_MeV: {state.kin_energy_MeV}")
```
Debug code:
```suggestion

```

ax3l commented on src/python/impactx/dashboard/jupyterApplication.py:
```diff
@@ -0,0 +1,28 @@
+        return application()
+
+
+"""
+from dashboard.jupyterApplication import JupyterMainApplication
+
+# Create new application instance
+app = JupyterMainApplication()
+
+# Let's start the server by waiting for its UI to be ready
+await app.ui.ready
+
+# Put the UI into the resulting cell
+app.ui
+"""
```
Is this a comment that we can remove? We have it in the docs now.
```suggestion
```

proy30 replied:
Yup, we can remove it now since it is in the docs.

ax3l replied:
Removed it in my last commit :)

ax3l committed:
Copyright Headers

proy30 committed:
Removed debug statements

proy30 committed:
Removed validation check for pos. number

proy30 committed:
Merge branch 'development' of github.com:proy30/impactx into development

ax3l commented on docs/source/usage/dashboard.rst:
```diff
@@ -0,0 +1,79 @@
+Dashboard
+=========
+
+This document provides instructions on how to launch the ImpactX Dashboard, a browser-based interface to ImpactX.
```
```suggestion
ImpactX Dashboard is a browser-based interface to ImpactX.
It provides a graphical interface to a subset of ImpactX functionality.

.. note::

   ImpactX Dashboard is provided as a preview and continues to be developed.
   Let us know in GitHub `discussions <https://github.com/ECP-WarpX/impactx/discussions>`__ and `issues <https://github.com/ECP-WarpX/impactx/issues>`__ how it works for you and what kind of workflows you would like to run in it.
```

ax3l commented on docs/source/usage/dashboard.rst:
```diff
@@ -0,0 +1,79 @@
+Launching the Dashboard
+-----------------------
+
+1. **Standalone browser application:**
```
```suggestion
The ImpactX Dashboard can be run in two modes, as a standalone browser application or inside a Jupyter notebook.

1. **Standalone browser application:**
```

ax3l committed:
Update Doc Lingo

ax3l commented on docs/source/usage/dashboard.rst:
```diff
@@ -0,0 +1,87 @@
+      app.ui
+
+
+Navigating the GUI
+------------------
+
+The GUI is designed with a user-friendly interface that includes multiple tabs and menu options:
```
```suggestion
Navigation
----------

The user-friendly interface includes multiple tabs and menu options, intended to be navigated from top to bottom:
```

ax3l committed:
Usage Navigation

ax3l commented on src/python/impactx/dashboard/Analyze/plotsMain.py:
```diff
@@ -0,0 +1,249 @@
+REDUCED_BEAM_DATA = os.path.join(DIAGS_DIR, "reduced_beam_characteristics.0.0")
+REF_PARTICLE_DATA = os.path.join(DIAGS_DIR, "ref_particle.0.0")
+
+DEFAULT_HEADERS = ["step", "s", "alpha_x", "alpha_y", "alpha_t"]
```
Better defaults :)
```suggestion
DEFAULT_HEADERS = ["s", "beta_x", "beta_y"]
```
Essentially shows the beam envelope over beamline distance.

ax3l commented on src/python/impactx/dashboard/Analyze/plot_PhaseSpaceProjections/phaseSpace.py:
```diff
@@ -0,0 +1,120 @@
+    # simulate
+    sim.evolve()
+
+    # check local particles
+    df = pc.to_df(local=True)
+    print(df)
+
+    # ensure the column heads are correctly labeled
+    assert df.columns.tolist() == [
+        "idcpu",
+        "position_x",
+        "position_y",
+        "position_t",
+        "momentum_x",
+        "momentum_y",
+        "momentum_t",
+        "qm",
+        "weighting",
+    ]
```
This is from a unit-test, does not belong here:
```suggestion
```

ax3l commented on src/python/impactx/dashboard/Analyze/plot_PhaseSpaceProjections/phaseSpace.py:
```diff
@@ -0,0 +1,120 @@
+    return base64.b64encode(buf.read()).decode("utf-8")
+
+
+def run_simulation(save_png=True):
```
Why does the `phaseSpace.py` analysis logic have a `run_simulation` function?
Should this be located somewhere else?

proy30 replied:
I was not sure how to retrieve the data for the phase space plot, nor did I find anything in the examples where the simulation and phase space plots were done in different functions, which is why I kept it as one.

ax3l replied:
What I mean is: the simulation control should be in its own file. Looks a bit odd to have the central simulation run logic located here in an analysis sub-folder.

Once the sim has evolved, you can return the `sim` object or so and plot from it?

proy30 replied:
Yes, this makes sense. I think the simulation script could be located in a folder called 'run'? I need to do a little testing with returning the sim object and plotting from that. Am I supposed to return 'sim' or 'pc'?

ax3l replied:
`pc` is only valid as long as the `sim` object is alive, so it might be more general to return `sim` and query it for post-processing. our garbage collection (should) know not to delete sim as long as pc is alive, but passing `sim` around will allow more operations.

when done, e.g., when starting new runs, finalize and delete the old `sim`.

ax3l commented on src/python/impactx/dashboard/Analyze/plot_PhaseSpaceProjections/phaseSpace.py:
```diff
@@ -0,0 +1,120 @@
+
+    sim.add_particles(bunch_charge_C, distr, npart)
+
+    assert pc.total_number_of_particles() == npart
+
```
This is from a unit test, pls remove:
```suggestion
```

ax3l commented on src/python/impactx/dashboard/Analyze/plot_PhaseSpaceProjections/phaseSpace.py:
```diff
@@ -0,0 +1,120 @@
+    ref.set_charge_qe(-1.0).set_mass_MeV(0.510998950).set_kin_energy_MeV(kin_energy_MeV)
+
+    #   particle bunch
+    # distr = distribution.Waterbag(
+    #     lambdaX=3.9984884770e-5,
+    #     lambdaY=3.9984884770e-5,
+    #     lambdaT=1.0e-3,
+    #     lambdaPx=2.6623538760e-5,
+    #     lambdaPy=2.6623538760e-5,
+    #     lambdaPt=2.0e-3,
+    #     muxpx=-0.846574929020762,
+    #     muypy=0.846574929020762,
+    #     mutpt=0.0,
+    # )
```
```suggestion
```

ax3l commented on src/python/impactx/dashboard/Analyze/plot_PhaseSpaceProjections/phaseSpace.py:
```diff
@@ -0,0 +1,120 @@
+    assert pc.total_number_of_particles() == npart
+
+    # init accelerator lattice
+    # fodo = [
+    #     elements.Drift(0.25),
+    #     elements.Quad(1.0, 1.0),
+    #     elements.Drift(0.5),
+    #     elements.Quad(1.0, -1.0),
+    #     elements.Drift(0.25),
+    # ]
```
```suggestion
```

ax3l commented on src/python/impactx/dashboard/Input/latticeConfiguration/latticeMain.py:
```diff
@@ -0,0 +1,445 @@
+    return parameter_input
+
+
+def save_lattice_elements():
```
Would it be more descriptive if we rename this function to `get_lattice_elements` or `read_lattice_elements` ?

ax3l commented on src/python/impactx/dashboard/Input/distributionParameters/twissParameters.txt:
```diff
@@ -0,0 +1,88 @@
+twiss_parameters = {
```
This file is currently unused, let us remove it?

ax3l commented on src/python/impactx/dashboard/Analyze/plotsMain.py:
```diff
@@ -0,0 +1,249 @@
+CURRENT_DIR = os.getcwd()
+DIAGS_DIR = os.path.join(CURRENT_DIR, "diags")
+
+REDUCED_BEAM_DATA = os.path.join(DIAGS_DIR, "reduced_beam_characteristics.0.0")
+REF_PARTICLE_DATA = os.path.join(DIAGS_DIR, "ref_particle.0.0")
```
You could use this to read those files:
```py
def read_file(file_pattern):
    for filename in glob.glob(file_pattern):
        df = pd.read_csv(filename, delimiter=r"\s+")
        if "step" not in df.columns:
            step = int(re.findall(r"[0-9]+", filename)[0])
            df["step"] = step
        yield df

def read_time_series(file_pattern):
    """Read in all CSV files from each MPI rank (and potentially OpenMP
    thread). Concatenate into one Pandas dataframe.

    Returns
    -------
    pandas.DataFrame
    """
    return pd.concat(
        read_file(file_pattern),
        axis=0,
        ignore_index=True,
    )  # .set_index('id')
```
and then
```suggestion
REDUCED_BEAM_DATA = read_time_series(
    DIAGS_DIR + "/reduced_beam_characteristics.*"
)
REF_PARTICLE_DATA = read_time_series(DIAGS_DIR + "/ref_particle.*")
```

proy30 replied:
I'm not too sure what the difference of having these new functions would do. Could you please explain? I have some other functions in analyzeFunctions that merge the dataframes into one

ax3l replied:
Ah I see. Great, that works.

Small update: note that these files end in .number or .number.othernumber depending on running with or without MPI. Write the globbing general enough to be able to read either naming scheme please

proy30 committed:
removed unused code

proy30 committed:
Adjusted plotting over s default columns

ax3l commented on src/python/impactx/dashboard/Analyze/plot_PhaseSpaceProjections/phaseSpace.py:
```diff
@@ -0,0 +1,120 @@
+
+    sim.particle_shape = state.particle_shape
+    sim.space_charge = False
+    sim.slice_step_diagnostics = False
```
```suggestion
    sim.slice_step_diagnostics = True
```

proy30 committed:
Set slice_step_diagnostics to True

proy30 committed:
Updated diags file paths to use globbing for flexible naming

proy30 committed:
Removed all print debug statements

proy30 committed:
Moved reading of diags filed to once simulation is ran

proy30 committed:
Renamed save_distribution/lattice functions

And removed functions calls when unneeded

proy30 committed:
renamed simulation variables

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 commented on src/python/impactx/dashboard/Analyze/plot_PhaseSpaceProjections/phaseSpace.py:
```diff
@@ -57,12 +55,12 @@ def run_simulation():
     ref = pc.ref_particle()
     ref.set_charge_qe(-1.0).set_mass_MeV(0.510998950).set_kin_energy_MeV(kin_energy_MeV)
 
-    distr = distribution_parameters()
```
@ax3l , Would renaming the variables like this work? Maybe this goes against naming conventions and there is a better name we can give for the variables?

ax3l replied:
@proy30 can you tell me again renamed from which to which name? Because of the double change, I cannot see anymore what `distr = distribution_parameters()` was renamed into.

proy30 replied:
'distr' changed to 'distribution' 
'fodo' changed to 'lattice_configuration'

ax3l replied:
ah, that works well here. The other names were a specific example setup.

proy30 committed:
Removed 'step' and 's' headers for user selection for plot_over_s

s is shown on x-axis by default

proy30 committed:
Added 'keV' to kin_energy units option

proy30 committed:
Merge branch 'development' of github.com:proy30/impactx into development

ax3l committed:
Docs: Developer Deps for Trame

ax3l commented on src/python/impactx/dashboard/Analyze/plot_PhaseSpaceProjections/phaseSpace.py:
```diff
@@ -0,0 +1,77 @@
+    return base64.b64encode(buf.read()).decode("utf-8")
+
+
+def run_simulation():
```
In a follow-up PR, let's move this out to separate run logic controls.

ax3l reviewed (approved):
Fantastic work, congratulations :sparkles:
